xrmocap/transform/keypoints3d/optim/fourdag_optimization.py

```diff
@@ -0,0 +1,635 @@
+# yapf: disable
+import copy
+import numpy as np
+from typing import Union
+
+from xrmocap.ops.triangulation.builder import BaseTriangulator
+from xrmocap.transform.keypoints3d.optim.fourdag_base_optimizer import (
+    FourDAGBaseOptimizer,
+)
+from xrmocap.utils.fourdag_utils import (
+    LIMB_INFO, rodrigues, rodrigues_jacobi, welsch,
+)
+
+# yapf: enable
+
+
+class LimbInfo():
+
+    def __init__(self, kps_convention) -> None:
+        """save limb information."""
+        self.kps_convention = kps_convention
+        self.boneLen = np.zeros(
+            LIMB_INFO[self.kps_convention]['n_kps'] - 1, dtype=np.float32)
+        self.boneCnt = np.zeros(
+            LIMB_INFO[self.kps_convention]['n_kps'] - 1, dtype=np.float32)
+        self.active = 0.0
+        self.shape_fixed = False
+        self.data = np.zeros(
+            3 + LIMB_INFO[self.kps_convention]['n_kps'] * 3 +
+            LIMB_INFO[self.kps_convention]['shape_size'],
+            dtype=np.float32)
+
+    def push_previous_bones(self, limb):
+        for kps_id in range(1, LIMB_INFO[self.kps_convention]['n_kps']):
+            prtIdx = LIMB_INFO[self.kps_convention]['kps_parent'][kps_id]
+            if limb[3, kps_id] > 0 and limb[3, prtIdx] > 0:
+                len = np.linalg.norm(limb[:, kps_id][:3] - limb[:, prtIdx][:3])
+                self.boneLen[kps_id - 1] = (self.boneCnt[kps_id - 1] *
+                                            self.boneLen[kps_id - 1] + len) / (
+                                                self.boneCnt[kps_id - 1] + 1)
+                self.boneCnt[kps_id - 1] += 1
+
+    def get_trans(self):
+        return self.data[:3]
+
+    def get_pose(self):
+        return self.data[3:3 + LIMB_INFO[self.kps_convention]['n_kps'] * 3]
+
+    def get_trans_pose(self):
+        return self.data[:3 + LIMB_INFO[self.kps_convention]['n_kps'] * 3]
+
+    def get_shape(self):
+        return self.data[-LIMB_INFO[self.kps_convention]['shape_size']:]
+
+
+class Term():
```

Comment:
As offline discussion, we may split `Term` into two classes for static config and dynamic params respectively, and adopt a more specific name.